Adding C# version history post

erikdietrich opened the pull request:
Per Bill Wagner's request over email, we're adding the C# Version History blog post from NDepend to the documentation.

erikdietrich committed:
Create csharp-version-history.md

erikdietrich committed:
Added the actual text

erikdietrich committed:
Updated TOC with my entry.

dnfclas commented:
@__erikdietrich__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

dnfclas commented:
__@erikdietrich__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

BillWagner committed:
first pass edit

1. Add Metadata
1. Change person
1. Formalize from blog post to docs style language.

BillWagner commented:
I made a pass at tightening the language to match the structure of docs.

@mairaw Do you want to take a look?

@erikdietrich I tried to stay honest to your original intent while following our guidelines. If any of the edits are a concern, let me know, or push an update.

erikdietrich commented:
@BillWagner Looks good to me -- no issues here.

mairaw committed:
Merge branch 'master' into master

mairaw commented:
Sorry, I've missed this earlier. I'll get this reviewed soon. Apologies on the delay.

I've just resolved the merge conflicts we had.

lextm commented:
If the content can include some reference to Visual J++ and its extensions of Java, then the history would be more naturally showing the transition to C#. The discontinuation of J++ and the birth of C# are associated with each other.

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+Oh, and speaking of iterators, C# Version 2 brought iterators.  To put it succinctly, this let you iterate through the items in List (or other Enumerable types) with a foreach loop.  Having this as a first class part of the language dramatically enhanced readability of the language and people&#39;s ability to reason about the code.
+
+And yet, C# continued to play a bit of catch up with Java.  Java had already released versions that included generics and iterators.  But that would soon change as the languages continued to evolve apart.
```
catch up -> catch-up

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+- [Expression trees](https://blogs.msdn.microsoft.com/charlie/2008/01/31/expression-tree-basics/)
+- [Extension methods](https://www.codeproject.com/Tips/709310/Extension-Method-In-Csharp)
+
+In retrospect, many of these features seem both inevitable and inseparable.  They all fit together so strategically.  It's generally thought that C# Version 3&#39;s killer feature was the query expression, also known as Linq ( **L** anguage **IN** tegrated **Q** uery).
```
Linq ( **L** anguage **IN** tegrated **Q** uery) -> Language-Integrated Query (LINQ)

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+In retrospect, many of these features seem both inevitable and inseparable.  They all fit together so strategically.  It's generally thought that C# Version 3&#39;s killer feature was the query expression, also known as Linq ( **L** anguage **IN** tegrated **Q** uery).
+
+A more nuanced view examines expression tress, lamba expressions and anonymous types as the foundation upon which LINQ is constructed.  But, in either case, C# 3 presented with a fairly revolutionary concept.  C# 3 had begun to lay the groundwork for turning C# into a hybrid OO-functional language.
```
typo: lamba

missing comma -> lambda expressions, and anonymous types

to use an acronym, you have to introduce it at first similar to the example I used before with LINQ

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+A more nuanced view examines expression tress, lamba expressions and anonymous types as the foundation upon which LINQ is constructed.  But, in either case, C# 3 presented with a fairly revolutionary concept.  C# 3 had begun to lay the groundwork for turning C# into a hybrid OO-functional language.
+
+Specifically, you could now write SQL-style, declarative queries to perform operations on collections, among other things.  Instead of writing a for loop to compute the average of a list of integers, you could now do that as simply as list.Average().  The combination of query expressions and extension methods made it look as though that list of ints had gotten a whole lot smarter.
```
keywords and such need to be inside backticks

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+## C# Version 4
+
+C# Version 4 would have had a difficult time living up to the groundbreaking status of version 3.  With version 3, C# had moved the language firmly out from the shadow of Java and into prominence.  The language was quickly becoming elegant
```
missing punctuation

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+C# Version 4 would have had a difficult time living up to the groundbreaking status of version 3.  With version 3, C# had moved the language firmly out from the shadow of Java and into prominence.  The language was quickly becoming elegant
+
+The next version did introduce some cool stuff, though.
```
cool stuff: very informal 

Suggestion: 
However, this version still introduced some interesting new features:

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+Embedded interop types alleviated a deployment pain.  Generic covariance and contravariance give you a lot of power, but they&#39;re a bit academic and probably most appreciated by framework and library authors.  Named and optional parameters let you eliminate a lot of method overloads and provide convenience.  But none of those are exactly paradigm altering.
+
+The major feature was the introduction of the dynamic keyword.  The `dynamic` introduced into C# Version 4 the ability to override the compiler on compile time typing.  By using the dynamic keyword, you can create constructs similar to dynamically typed languages like JavaScript.  You can create a dynamic x = &quot;a string&quot; and then add six to it, leaving it up to the runtime to sort out wh ...
```
compile time -> compile-time

backticks around the inline code 

what on should -> what should

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+## C# Version 5
+
+C# Version 5 wasa very focused version of the language.  Nearly all of their effort for that version into another pretty groundbreaking language concept.  Here is the major features list.
```
wasa -> was a 

remove pretty

Second sentence is missing a verb. It's probably went?

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+The caller info attribute is pretty cool.  It lets you easily retrieve information about the context in which you&#39;re running without resorting to a ton of boilerplate reflection code.  Many developers love this feature.
+
+But async and await are the real stars of this release.  When this came out in 2012, C# changed the game again by baking asynchrony into the language as a first class participant.  If you&#39;ve ever dealt with long running operations and the implementation of webs of callbacks, you probably loved this language feature.
```
first class -> first-class

When this -> When these *features*?

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+## C# Version 7
+
+Finally, we arrive at C# version 7.  This has some evolutionary and cool stuff in the vein of C# 6, but without the compiler as a service.  Here are some of the new features.
```
This has -> This version has

end with a colon to introduce the list of features

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+But if you looked back on C# 1.0 now, you&#39;d find yourself a little dizzy.  It lacked the built in async capabilities and some of the slick functionality around generics that we take for granted.  As a matter of fact, it lacked generics altogether.  And [Linq](https://msdn.microsoft.com/en-us/library/bb308959.aspx)?  Nope.  That would take some years to come out.
+
+C# version 1 looked pretty stripped of features, compared to today.  You&#39;d find yourself writing some verbose code.  But yet, you have to start somewhere.
```
pretty stripped -> stripped

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+- [Generic covariant and contravariant](https://docs.microsoft.com/en-us/dotnet/standard/generics/covariance-and-contravariance)
+- [Embedded interop types](https://stackoverflow.com/questions/20514240/whats-the-difference-setting-embed-interop-types-true-and-false-in-visual-studi)
+
+Embedded interop types alleviated a deployment pain.  Generic covariance and contravariance give you a lot of power, but they&#39;re a bit academic and probably most appreciated by framework and library authors.  Named and optional parameters let you eliminate a lot of method overloads and provide convenience.  But none of those are exactly paradigm altering.
```
a lot of methods -> many methods

But none of those *features*?

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+## C# Version 6
+
+With versions 3 and 5, C# had added some pretty impressive stuff in an OO language.  (Version 2 did as well, but it was fast following Java with those language features.)  With version 6, it would go away from doing a dominant killer feature and instead release a lot of features that delighted users of the language.  Here are some of them.
```
remove pretty

replace stuff with something less informal

a lot of features -> many features

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+With versions 3 and 5, C# had added some pretty impressive stuff in an OO language.  (Version 2 did as well, but it was fast following Java with those language features.)  With version 6, it would go away from doing a dominant killer feature and instead release a lot of features that delighted users of the language.  Here are some of them.
+
+- [Static imports (a la Java)](http://geekswithblogs.net/BlackRabbitCoder/archive/2015/04/16/c.net-little-wonders-static-using-statements-in-c-6.aspx)
```
a la Java -> we avoid non english terms in our topics. Perhaps replace it with like? or remove that completely?

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+Taken individually, these are all cool language features.  But if you look at them altogether, you see an interesting pattern.  In this version, C# eliminated language boilerplate to make code more terse and readable.  So for fans of clean, simple code, this language version was a huge win.
+
+Oh, and they did do one other thing along with this version, though it&#39;s not a traditional language feature, per se.  They released [Roslyn the compiler as a service](https://github.com/dotnet/roslyn).  C# now uses C# to build C#, and they let you use the compiler as part of your programming efforts.
```
remove Oh

they did do -> they did

per se -> in itself?

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+- [Expanded expression bodied members](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#more-expression-bodied-members)
+- [Ref locals and returns](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#ref-locals-and-returns)
+
+All of these offer cool new capabilities for developers and the opportunity to write even cleaner code than ever.  A highlight is condensing the declaration of variables to use with the `out` keyword and by allowing multiple return values via tuple.
```
All of these -> All of these features

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+ms.author: wiwagn
+ms.date: 2017/08/09
+ms.topic: article
+ms.prod: visual-studio-dev-15
```
ms.prod should be .net

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+keywords: C#, .NET, .NET Core, What's New, C# History
+author: erikdietrich
+ms.author: wiwagn
+ms.date: 2017/08/09
```
date format is mm/dd/yyyy. otherwise weird things happen 😄 

![image](https://user-images.githubusercontent.com/12971179/30261560-18c09012-9682-11e7-8119-ed6ba2c1dd48.png)

mairaw reviewed (changes requested):
Really nice @erikdietrich! Thanks so much.

@BillWagner please let me know if I went too far on trying to bring this in to our style. I'm happy to help to incorporate my suggestions if needed. Please let me know.

BillWagner commented:
@mairaw Thanks for the review.  I'll update it today.

BillWagner commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+What did the language look like in its earliest incarnations?  And how has it evolved in the years since?
+
+## C# Version 1
```
updated to 1.0, etc. Now that we do point releases.

BillWagner committed:
Update csharp-version-history.md

BillWagner commented:
@mairaw Updates in place.

mairaw commented:
I'll go through this PR now but build has some warnings. Wanna take a look @BillWagner?

BillWagner committed:
fix build warnings on links.

dnfclas commented:
@erikdietrich,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

mairaw commented on docs/csharp/whats-new/csharp-version-history.md:
```diff
@@ -0,0 +1,124 @@
+
+With versions 3.0 and 5.0, C# had added some impressive features in an Object Oriented language. (Version 2.0 did as well, but it was fast following Java with those language features.) With version 6.0, it would go away from doing a dominant killer feature and instead release many features that delighted users of the language. Here are some of them.
+
+- [Static imports](http://geekswithblogs.net/BlackRabbitCoder/archive/2015/04/16/c.net-little-wonders-static-using-statements-in-c-6.aspx)
```
link to https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-static instead?

mairaw commented:
@BillWagner let me know if you'd like me to work on these updates for you. Thanks!

BillWagner committed:
respond to feedback

BillWagner commented:
@mairaw I just edited and responded to all the feedback.

mairaw committed:
fixed build warnings

mairaw committed:
fixed version capitalization
